modules/core/include/opencv2/core/utility.hpp

```diff
@@ -674,8 +676,39 @@ void Mat::forEach_impl(const Functor& operation) {
 
 /////////////////////////// Synchronization Primitives ///////////////////////////////
 
+class CV_EXPORTS CEEMutex
+{
+  public:
+    CEEMutex(void);
+    ~CEEMutex();
+	  CEEMutex(const CEEMutex&) = delete;
+	  CEEMutex& operator=(const CEEMutex&) = delete;
+  public:
+    void lock(void);
+    void unlock(void);
+	  bool try_lock(void) noexcept;
+  private:
+    void* impl;
+};
+
+class CV_EXPORTS CEELockGuard
+{
+  public:
+    explicit CEELockGuard(CEEMutex& _Mtx);
+    ~CEELockGuard();
+	  CEELockGuard(const CEELockGuard&) = delete;
+	  CEELockGuard& operator=(const CEELockGuard&) = delete;
+  private:
+    void* impl;
+};
+
+#if !defined(_M_CEE)
 typedef std::recursive_mutex Mutex;
 typedef std::lock_guard<cv::Mutex> AutoLock;
+#else
+typedef CEEMutex Mutex;
```

Comment:
Can we just put custom definitions of Mutex and AutoLock here instead of introducing proxy classes CEEMutex and CEELockGuard?

Reply:
I don't think so. CEEMutex and CEELockGuard must be compiled anyway, so that the dll built without /clr will still work when linked to a /clr program including opencv headers

Reply:
What API requires Mutex/AutoLock emulation?

Reply:
KMeansDistanceComputer

Reply:
Looks like, mutex is misused there. Removed: #12660

Reply:
Indeed, when #12660 is commited, cv::Mutex can just be made unavailable from /clr builds (and CEEMutex/CEELockGuard can be removed)
The only drawback is that under a /clr build, the developer won't be able to just copy/paste valid code using cv::Mutex or cv::AuLock, and will have to rely on a custom implementation of his own

Reply:
I believe CLR assumes using of own mechanics for synchronization instead of raw Mutex.